src/mpid/ch4/shm/posix/eager/iqueue/iqueue_send.h

```diff
@@ -71,59 +73,45 @@ MPIDI_POSIX_eager_send(int grank,
 
     /* Figure out the capacity of each cell */
     capacity = MPIDI_POSIX_EAGER_IQUEUE_CELL_CAPACITY(transport);
+
     available = capacity;
 
     cell->from = MPIDI_POSIX_global.my_local_rank;
 
     /* If this is the beginning of the message, mark it as the head. Otherwise it will be the
      * tail. */
-    if (*msg_hdr) {
+    cell->payload_size = 0;
+    if (am_hdr) {
+        MPI_Aint resized_am_hdr_sz = MPIDI_POSIX_RESIZE_TO_MAX_ALIGN(am_hdr_sz);
         cell->am_header = **msg_hdr;
-        *msg_hdr = NULL;        /* completed header copy */
         cell->type = MPIDI_POSIX_EAGER_IQUEUE_CELL_TYPE_HDR;
+        /* send am_hdr if this is the first segment */
+        MPIR_Typerep_copy(payload, am_hdr, am_hdr_sz);
+        /* make sure the data region starts at the boundary of MAX_ALIGNMENT */
+        payload = payload + resized_am_hdr_sz;
+        cell->payload_size += resized_am_hdr_sz;
+        cell->am_header.am_hdr_sz = resized_am_hdr_sz;
+        available -= cell->am_header.am_hdr_sz;
     } else {
         cell->type = MPIDI_POSIX_EAGER_IQUEUE_CELL_TYPE_DATA;
     }
 
-    /* Pack the data into the cells */
-    iov_done = 0;
-    for (i = 0; i < *iov_num; i++) {
-        /* Optimize for the case where the message will fit into the cell. */
-        if (likely(available >= (*iov)[i].iov_len)) {
-            MPIR_Typerep_copy(payload, (*iov)[i].iov_base, (*iov)[i].iov_len);
-
-            payload += (*iov)[i].iov_len;
-            available -= (*iov)[i].iov_len;
-
-            iov_done++;
+    if (data_sz_left) {
+        send_size = MPL_MIN(available, data_sz_left);
+        if (need_packing) {
+            MPIR_Typerep_pack(buf, count, datatype, offset, payload, send_size, &packed_size);
+            send_size = packed_size;
         } else {
-            /* If the message won't fit, put as much as we can and update the iovec for the next
-             * time around. */
-            MPIR_Typerep_copy(payload, (*iov)[i].iov_base, available);
-
-            (*iov)[i].iov_base = (char *) (*iov)[i].iov_base + available;
-            (*iov)[i].iov_len -= available;
-
-            available = 0;
-
-            break;
+            MPIDI_Datatype_check_lb(datatype, dt_true_lb);
+            MPIR_Typerep_copy(payload, (char *) buf + dt_true_lb + offset, send_size);
         }
     }
```

Comment:
You were saying `MPIR_Typerep_copy` has two less branches than `MPIR_Typerep_pack`, but aren't we adding the branches here? line 99, 101, and line 105?

Reply:
This is these two are the same amount of branches. The old has two if statements, the new one is a switch statement which has two branches.

This is also unrelated to the branching between copy and pack.

Reply:
@hzhou have you spotted any other problem?